### PR TITLE
Add support for binary numbers

### PR DESCRIPTION
**Pull request by uxcn**
I submitted a version of this patch to the Vim mailing list, which I'm still hoping to get feedback on.  I'd like to see if it might also be useful for NeoVim.

Since a number of languages have had or recently added support for binary encoded numbers  (Python, Ruby, Java, C/C++, etc...), it seemed reasonable to add native support in Vim/NeoVim.  Obviously there has already been support for octal and hexadecimal.

A quick summary of what's added...
- bin nrformat (default)
- prefix parsing (0b, 0B)
- value sort (:sort b)
- increment/decrement (^A, ^X)
- str2nr (base 2)
- binary encodings in VimL (e.g. 0b101)
- other miscellany (spell check, documentation, tests, and so on...)

I tried to avoid changing anything that wasn't strictly related to binary support.  If there are any issues with the patch, please let me know.  Note, there are a couple slightly thorny scenarios which I added tests for.

### CODE REVIEW

File: src/nvim/charset.c [hunk 1]
```diff
@@ -1454,6 +1454,20 @@ char_u* skipdigits(char_u *q)
   return p;
 }
 
+/// skip over binary digits
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Sentences must start with a capital letter.

File: src/nvim/charset.c [hunk 2]
```diff
@@ -1454,6 +1454,20 @@ char_u* skipdigits(char_u *q)
...
+///
+/// @param q
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Missing documentation.

File: src/nvim/charset.c [hunk 3]
```diff
@@ -1454,6 +1454,20 @@ char_u* skipdigits(char_u *q)
...
+///
+/// @return Pointer to the character after the skipped digits.
+char_u* skipbin(char_u *q)
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Do not use `char_u` in new functions dealing with strings. This needs to be `const char *skipbin(const char *q)`.

File: src/nvim/charset.c [hunk 4]
```diff
@@ -1454,6 +1454,20 @@ char_u* skipdigits(char_u *q)
...
+{
+    char_u	*p = q;
+
+    while (ascii_isbdigit(*p))	/* skip to next non-digit */
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Missing `{` around cycle body. `/*` comment should use `//`. Should have two spaces before comment start.

File: src/nvim/charset.c [hunk 5]
```diff
@@ -1454,6 +1454,20 @@ char_u* skipdigits(char_u *q)
...
+	++p;
```
Comment (ZyX-I) [inline src/nvim/charset.c]: According to the style guide increment/decrement statements should use post-increments. This rule is only not applicable if increment/decrement is a part of the expression.

File: src/nvim/charset.c [hunk 6]
```diff
@@ -1485,6 +1499,20 @@ char_u* skiptodigit(char_u *q)
   return p;
 }
 
+/// skip to binary character (or NUL after the string)
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Capital letter

File: src/nvim/charset.c [hunk 7]
```diff
@@ -1485,6 +1499,20 @@ char_u* skiptodigit(char_u *q)
...
+///
+/// @param q
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Documentation

File: src/nvim/charset.c (hunk 3 above)
Comment (ZyX-I) [inline src/nvim/charset.c]: Also FUNC_ATTR_NONNULL_ALL and FUNC_ATTR_PURE

File: src/nvim/charset.c [hunk 8]
```diff
@@ -1454,6 +1454,20 @@ char_u* skipdigits(char_u *q)
...
 }
 
+/// skip over binary digits
+///
+/// @param q
+///
+/// @return Pointer to the character after the skipped digits.
+char_u* skipbin(char_u *q)
+{
+    char_u	*p = q;
```
Comment (ZyX-I) [inline src/nvim/charset.c]: BTW, this is not needed since you can work with argument directly.

File: src/nvim/charset.c [hunk 9]
```diff
@@ -1485,6 +1499,20 @@ char_u* skiptodigit(char_u *q)
   return p;
 }
 
+/// skip to binary character (or NUL after the string)
+///
+/// @param q
+///
+/// @return Pointer to the binary character or (NUL after the string).
+char_u* skiptobin(char_u *q)
```
Comment (ZyX-I) [inline src/nvim/charset.c]: `const char *skiptobin(const char *q) FUNC_ATTR_NONNULL_ALL FUNC_ATTR_NONNULL_RET FUNC_ATTR_PURE`.

File: src/nvim/charset.c (hunk 3 above)
Comment (ZyX-I) [inline src/nvim/charset.c]: And FUNC_ATTR_NONNULL_RET

File: src/nvim/charset.c [hunk 10]
```diff
@@ -1788,28 +1825,38 @@ void vim_str2nr(char_u *start, int *hexp, int *len, int dooct, int dohex,
...
     }
-  } else if ((hex != 0) || (dohex > 1)) {
+  } else if (pre != 0 || dohex > 1) {
     // hex
+    if (pre != 0)
+      n += 2; // skip over "0x"
     while (ascii_isxdigit(*ptr)) {
       un = 16 * un + (unsigned long)hex2nr(*ptr);
-      ptr++;
+      ++ptr;
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Revert

File: src/nvim/charset.c [hunk 11]
```diff
@@ -1788,28 +1825,38 @@ void vim_str2nr(char_u *start, int *hexp, int *len, int dooct, int dohex,
...
     }
   } else {
     // decimal
     while (ascii_isdigit(*ptr)) {
       un = 10 * un + (unsigned long)(*ptr - '0');
-      ptr++;
+      ++ptr;
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Revert

File: src/nvim/eval.c [hunk 12]
```diff
@@ -1147,7 +1147,7 @@ call_vim_function (
       len = 0;
     else
       /* Recognize a number argument, the others must be strings. */
-      vim_str2nr(argv[i], NULL, &len, TRUE, TRUE, &n, NULL);
```
Comment (ZyX-I) [inline src/nvim/eval.c]: `//`-style comment and braces around both `if` and `else`. Since you are changing some line linter will require you to fix lint errors on adjacent lines.

File: src/nvim/eval.c [hunk 13]
```diff
@@ -4125,7 +4125,7 @@ eval7 (
         rettv->vval.v_float = f;
       }
     } else {
-      vim_str2nr(*arg, NULL, &len, TRUE, TRUE, &n, NULL);
+      vim_str2nr(*arg, NULL, &len, TRUE, TRUE, TRUE, &n, NULL);
```
Comment (ZyX-I) [inline src/nvim/eval.c]: You must not use TRUE and FALSE constants.

File: src/nvim/eval.c [hunk 14]
```diff
@@ -1147,7 +1147,7 @@ call_vim_function (
       len = 0;
     else
       /* Recognize a number argument, the others must be strings. */
-      vim_str2nr(argv[i], NULL, &len, TRUE, TRUE, &n, NULL);
+      vim_str2nr(argv[i], NULL, &len, TRUE, TRUE, TRUE, &n, NULL);
```
Comment (ZyX-I) [inline src/nvim/eval.c]: true

File: src/nvim/eval.c [hunk 15]
```diff
@@ -15804,7 +15804,7 @@ static void f_str2nr(typval_T *argvars, typval_T *rettv)
   p = skipwhite(get_tv_string(&argvars[0]));
   if (*p == '+')
     p = skipwhite(p + 1);
-  vim_str2nr(p, NULL, NULL, base == 8 ? 2 : 0, base == 16 ? 2 : 0, &n, NULL);
+  vim_str2nr(p, NULL, NULL, base == 2 ? 2 : 0, base == 8 ? 2 : 0, base == 16 ? 2 : 0, &n, NULL);
```
Comment (ZyX-I) [inline src/nvim/eval.c]: Too long line

File: src/nvim/eval.c [hunk 16]
```diff
@@ -18086,7 +18086,7 @@ long get_tv_number_chk(typval_T *varp, int *denote)
   case VAR_STRING:
     if (varp->vval.v_string != NULL)
       vim_str2nr(varp->vval.v_string, NULL, NULL,
-          TRUE, TRUE, &n, NULL);
+          TRUE, TRUE, TRUE, &n, NULL);
```
Comment (ZyX-I) [inline src/nvim/eval.c]: true

Comment (ZyX-I) [inline src/nvim/eval.c]: First true should be aligned with varp

File: src/nvim/ex_cmds.c [hunk 17]
```diff
@@ -348,6 +348,7 @@ void ex_sort(exarg_T *eap)
   long deleted;
   colnr_T start_col;
   colnr_T end_col;
+  int sort_bin;                         /* sort on bin number */
```
Comment (ZyX-I) [inline src/nvim/ex_cmds.c]: // comment, same for next comments (remember that linter requires changing adjacent lines). Also comment should not be that far to the right.

File: src/nvim/ex_cmds.c [hunk 18]
```diff
@@ -373,6 +374,8 @@ void ex_sort(exarg_T *eap)
       sort_rx = TRUE;
     else if (*p == 'n')
       sort_nr = 2;
+    else if (*p == 'b')
+      sort_bin = 2;
```
Comment (ZyX-I) [inline src/nvim/ex_cmds.c]: Braces, here and in all of the other lines of this if. Though I would really convert this to `switch`

File: src/nvim/ex_cmds.c [hunk 19]
```diff
@@ -410,14 +413,14 @@ void ex_sort(exarg_T *eap)
     }
   }
 
-  /* Can only have one of 'n', 'o' and 'x'. */
-  if (sort_nr + sort_oct + sort_hex > 2) {
+  /* Can only have one of 'n', 'b', 'o' and 'x'. */
```
Comment (ZyX-I) [inline src/nvim/ex_cmds.c]: //

File: src/nvim/ex_cmds.c [hunk 20]
```diff
@@ -454,6 +457,8 @@ void ex_sort(exarg_T *eap)
       p = s + start_col;
       if (sort_hex)
         s = skiptohex(p);
+      else if (sort_bin)
```
Comment (ZyX-I) [inline src/nvim/ex_cmds.c]: braces

File: src/nvim/ex_cmds.c [hunk 21]
```diff
@@ -462,7 +467,7 @@ void ex_sort(exarg_T *eap)
         /* empty line should sort before any number */
         nrs[lnum - eap->line1].start_col_nr = -MAXLNUM;
       else
-        vim_str2nr(s, NULL, NULL, sort_oct, sort_hex,
+        vim_str2nr(s, NULL, NULL, sort_bin, sort_oct, sort_hex,
             &nrs[lnum - eap->line1].start_col_nr, NULL);
```
Comment (ZyX-I) [inline src/nvim/ex_cmds.c]: Alignment

File: src/nvim/ex_getln.c [hunk 22]
```diff
@@ -4780,7 +4780,7 @@ int get_list_range(char_u **str, int *num1, int *num2)
 
   *str = skipwhite(*str);
   if (**str == '-' || ascii_isdigit(**str)) {  /* parse "from" part of range */
-    vim_str2nr(*str, NULL, &len, FALSE, FALSE, &num, NULL);
+    vim_str2nr(*str, NULL, &len, FALSE, FALSE, FALSE, &num, NULL);
```
Comment (ZyX-I) [inline src/nvim/ex_getln.c]: false

File: src/nvim/ex_getln.c [hunk 23]
```diff
@@ -4788,7 +4788,7 @@ int get_list_range(char_u **str, int *num1, int *num2)
   *str = skipwhite(*str);
   if (**str == ',') {                   /* parse "to" part of range */
     *str = skipwhite(*str + 1);
-    vim_str2nr(*str, NULL, &len, FALSE, FALSE, &num, NULL);
+    vim_str2nr(*str, NULL, &len, FALSE, FALSE, FALSE, &num, NULL);
```
Comment (ZyX-I) [inline src/nvim/ex_getln.c]: false

File: runtime/doc/change.txt [hunk 24]
```diff
@@ -386,6 +388,10 @@ octal and hexadecimal numbers and alphabetic characters.  This depends on the
   under or after the cursor.  This is useful to make lists with an alphabetic
   index.
 
+For decimals a leading negative sign is considered for incrementing/
```
Comment (ZyX-I) [inline runtime/doc/change.txt]: incrementing or decrementing. Slash could be fine replacement for “or” on one line, but not here.

File: src/nvim/ascii.h [hunk 25]
```diff
@@ -122,6 +122,14 @@ static inline bool ascii_isxdigit(int c)
          || (c >= 'A' && c <= 'F');
 }
 
+/// Checks if `c` is a binary digit, that is, 0-1.
+///
+/// @see {ascii_isdigit}
+static inline bool ascii_isbdigit(int c)
```
Comment (ZyX-I) [inline src/nvim/ascii.h]: FUNC_ATTR_CONST

Comment (ZyX-I) [inline src/nvim/ascii.h]: But not here. You are missing declaration below line 91.

**Comment by uxcn** [comment]
Awesome, thanks @ZyX-I.  I noticed some additional style issues in the functions I modified.  Should I leave those be?  Should I fix them in a separate commit?

**Comment by ZyX-I** [comment]
@uxcn If they are reported by the linter (`make lint`) then yes. If they are not then this is out of the scope of the PR.

**Comment by ZyX-I** [comment]
Separate or not commit depends on how massive change is. Massive style changes better be placed in a separate commit, local are fine in related commit. It is also OK to create one commit which fixes all style issues, local or not.

**Comment by ZyX-I** [comment]
Though this is how I write commits. Do not know the official policy (most likely this is something like “be sane and respect reviewers, that’s it”).

File: src/nvim/charset.c [hunk 26]
```diff
@@ -1720,33 +1748,38 @@ int vim_isblankline(char_u *lbuf)
...
+/// @param prep Returns type of number 0 = decimal, 'x' or 'X' is hex,
+//         '0' = octal, 'b' or 'B' is bin
 /// @param len Returns the detected length of number.
+/// @param dobin recognize binary number
 /// @param dooct recognize octal number
 /// @param dohex recognize hex number
 /// @param nptr Returns the signed result.
 /// @param unptr Returns the unsigned result.
-void vim_str2nr(char_u *start, int *hexp, int *len, int dooct, int dohex,
+void vim_str2nr(char_u *start, int *prep, int *len, int dobin, int dooct, int dohex,
```
Comment (justinmk) [inline src/nvim/charset.c]: Bram suggested coallescing these booleans to an enum. Do you plan to do that?

**Comment by justinmk** [comment]
Thanks @uxcn! Glancing at the code it looks good to me, with attention given to details like comments/documentation/tests.

Does the work here help any with converting hex representation to binary? I've often thought that would be useful for reading bit flags...

You can ignore the mingw build, looks like another broken download (similar to the luajit.org issue fixed yesterday).

File: runtime/doc/change.txt [hunk 27]
```diff
@@ -371,8 +371,10 @@ CTRL-X			Subtract [count] from the number or alphabetic
 			character at or after the cursor.
 
 The CTRL-A and CTRL-X commands work for (signed) decimal numbers, unsigned
-octal and hexadecimal numbers and alphabetic characters.  This depends on the
-'nrformats' option.
+binary and octal and hexadecimal numbers and alphabetic characters.  This
```
Comment (justinmk) [inline runtime/doc/change.txt]: binary/octal/hexadecimal numbers, and alphabetic characters

File: runtime/doc/change.txt [hunk 28]
```diff
@@ -371,8 +371,10 @@ CTRL-X			Subtract [count] from the number or alphabetic
...
+depends on the 'nrformats' option.
+- When 'nrformats' includes "bin", Vim considers numbers starting with '0b' or
+  '0B' are binary.
```
Comment (justinmk) [inline runtime/doc/change.txt]: are => as

File: test/functional/legacy/057_sort_spec.lua [hunk 29]
```diff
@@ -600,4 +600,39 @@ describe(':sort', function()
     eq('Vim(sort):E474: Invalid argument', eval('tmpvar'))
     expect(text)
   end)
+  
+  it('binary', function()
+    insert([[
+    0b111000
```
Comment (justinmk) [inline test/functional/legacy/057_sort_spec.lua]: A `0x`-prefixed value like `0x10b00` would be good to include here.

File: src/nvim/charset.c [hunk 30]
```diff
@@ -1720,33 +1748,38 @@ int vim_isblankline(char_u *lbuf)
...
 
 /// Convert a string into a long and/or unsigned long, taking care of
-/// hexadecimal and octal numbers.  Accepts a '-' sign.
-/// If "hexp" is not NULL, returns a flag to indicate the type of the number:
+/// hexadecimal, octal and binary numbers.  Accepts a '-' sign.
+/// If "prep" is not NULL, returns a flag to indicate the type of the number:
 ///   0      decimal
 ///   '0'    octal
+///   'B'	   bin
+///   'b'	   bin
```
Comment (justinmk) [inline src/nvim/charset.c]: align with the other items in this list

**Comment by uxcn** [comment]
Unfortunately the patch doesn't add total support for converting bases (e.g. hex2bin), but hopefully it's an okay step toward ultimately supporting it.  I have a VimL [plugin](https://github.com/uxcn/vim-x2x) I've used for generic conversions for a while (including arbitrary precision) in case anybody can use it now.  I was considering trying to submit another PR for something similar as time permitted, but I'm not sure where it fits with priorities.

I like the idea of a `%b` modifier for `printf` regardless.  I know I could personally use it, and I think some libc implementations even support it.  Unless someone tells me not to, I'll add it to the PR.

I noticed the MinGW build for travis.  I'm not too familiar with the code for the build, but if I see something I can fix, I'll add another PR.

File: src/nvim/charset.c (hunk 26 above)
Comment (uxcn) [inline src/nvim/charset.c]: I agree with Bram, but I wanted to avoid changing the semantics in the code at least for this PR.  Changing APIs always worries me.  There were a couple other things I noticed though.

I think a bitmask of *do* parse bases could replace the three *doX* bases, and then for the *doX* > 1, add an explicit base parameter.

Comment (ZyX-I) [inline src/nvim/charset.c]: @uxcn I do not think this is correct concern. By adding a new argument you already did change the API, and semantics cannot be changed by converting three boolean options into a flag. Semantics would change if you converted three booleans into a `struct { char *prefix; int base; }[]`: to generalize the function it is logical to make it be called like

    vim_str2nr("0b10110", &type, &len, (Bases[]) { { "0x", 16 }, { "0", 8 }, { "0b", 2 }, { NULL, 10 }, { NULL, 0 } }, &n, &un)

: change makes sense, but here you supply prefixes to try (determining order) and thus change function that supports fixed formats defined internally to a function that accepts rather broad range of formats defined externally.

Comment (uxcn) [inline src/nvim/charset.c]: Making `vim_str2nr` completely generic would be a much bigger change.  The function is pretty tightly coupled to base representations.  For example, pattern matching is done explicitly.  Patterns can also overlap (e.g. `0x0b01` could match as `0b01` or `0x0b01`).  There are other places in code that rely on explicit representations for numbers as well.  Another question might also be, should multiple prefixes be allowed (e.g. `0x` and `0h` or `h`)?

Comment (ZyX-I) [inline src/nvim/charset.c]: @uxcn It was an illustration to what I think *does* change the semantics, opposed to converting three booleans into a flag option which does not.

Comment (uxcn) [inline src/nvim/charset.c]: They actually have `int` semantics for now (1: maybe/prefix, 2: always/no-prefix).  The `ex_sort` function is roughly the same.

Comment (ZyX-I) [inline src/nvim/charset.c]: @uxcn This only changes how many bits one option will occupy in new flags variable.

**Comment by justinmk** [comment]
> I like the idea of a %b modifier for printf regardless. I know I could personally use it, and I think some libc implementations even support it. Unless someone tells me not to, I'll add it to the PR.

sounds good--separate commit?

**Comment by uxcn** [comment]
> sounds good--separate commit?

I was planning on one for the features and another for lint.  I can do a separate commit for `%b` though.

Is a force rebase an issue for a PR?

> > should multiple prefixes be allowed (e.g. 0x and 0h or h)?

> I would say no.

I would tend to agree.

**Comment by justinmk** [comment]
@uxcn rebase is no problem

**Comment by justinmk** [comment]
@uxcn Is this ready? Did Bram make any changes to your patch that should be added here?

**Comment by ZyX-I** [comment]
@justinmk My comments are still applicable.

**Comment by justinmk** [comment]
@ZyX-I Yes, I mean after the comments are addressed. Though the discussion about `vim_str2nr` is out of scope.

**Comment by uxcn** [comment]
I haven't heard anything back on the mailing list after Bram initially commented.  I might try following up again since it's after the holidays.  I would want to try to merge the `printf` stuff back, which is hopefully okay.

Otherwise, the code and lint fixes are done and I'll be updating the PR later tonight.  I have a few questions, but I'm going to update the PR first.

**Comment by justinmk** [comment]
@uxcn Just to be clear, Bram already merged your patch into Vim. Often he makes changes without mentioning it so I wanted to make sure those changes are included here.

**Comment by uxcn** [comment]
@justinmk Thanks, I actually had no idea he already merged it.  I'll go through and check to see what's necessary for the `printf` stuff tonight.

**Comment by uxcn** [comment]
The latest commits incorporate all of the comments as well as the `printf` functionality.  I still had a few questions if anyone could help me out though.

The linter marks `long long` for the `ll` modifier in `vim_vsnprintf` as an error, but I'm hesitant to change it since `long long` might not be the same on some platforms.  Obviously `printf` semantics have preferred matching the non-explicit width types for a while.  Is there a work-around?

Also, I'm wondering if there's a better way to [test](https://github.com/uxcn/neovim/blob/binary/test/functional/viml/printf_spec.lua) `printf`.  The test I wrote relies on whitespace and control characters, which seems a bit flaky.  Is there a better way to test VimL functions?

Finally I was just curious... Is there a reason `vim_vsnprintf` and `snprintf` are both used in the code?

**Comment by uxcn** [comment]
Also, it looks like one of the builds had an issue setting up python3 pip.  I think swapping `python3.3` in `.ci/before_install.sh` for `python3` might fix it, but I'm not entirely sure what the python environments look like with travis.  I do remember I've had issues in the past though.

**Comment by ZyX-I** [comment]
@uxcn Regarding `long long`: if it was not printf I would suggest using `uint64_t`. But here you need to use `// NOLINT (category)`.

Regarding test: if you have any options do *not* write legacy-style tests. Definitely do not do this in `test/functional/viml`, legacy tests are to be placed in `test/functional/legacy`. Put test for `printf` into `test/functional/eval`. You can check examples there, but none of them are using new `helpers.funcs` and your test is ideal for writing things like

 ```Lua
local helpers = require('test.functional.helpers')
local clear = helpers.clear
local eq = helpers.eq
local funcs = helpers.funcs
local exc_exec = helpers.exc_exec

describe('printf()', function()
  it('works with %b', function()
    eq('1', funcs.printf('%b', 1))
  end)
  it('accepts width modifier with %b', function()
    eq('  1', funcs.printf('%3b', 1))
  end)
  it('accepts prefix modifier with %b', function()
    eq('0b1', funcs.printf('%#b', 1))
  end)
  it('writes capital B with %B', function()
    eq('0B1', funcs.printf('%#B', 1))
  end)  -- By the way, I find this rather useless. Unlike hexadecimal there are no letter-digits.
  it('accepts prefix, zero-fill and width modifiers with %b', function()
    eq('0b001', funcs.printf('%#05b', 1))
  end)
  it('accepts prefix and width modifiers with %b', function()
    eq('  0b1', funcs.printf('%#5b', 1))
  end)
  it('does not write prefix for zero with prefix and width modifier used with %b', function()
    eq('    0', funcs.printf('%#5b', 0))
  end)
  it('accepts precision modifier with %b', function()
    eq('00000', funcs.printf('%.5b', 0))
  end)
  it('accepts all modifiers with %b at once', function()
    -- zero-fill modifier is ignored when used with left-align
    -- force-sign and add-blank are ignored
    -- use-grouping-characters modifier is ignored always
    eq('0b00011   ', funcs.printf('% \'+#0-10.5b', 3))
  end)
  it('converts strings to numbers with %b', function()
    eq('1111', funcs.printf('%b', '0xF'))
  end)
  it('errors out when %b modifier is used for a list', function()
    eq('Vim(call):E745:Using a List as a Number', exc_exec('call printf("%b", [])'))
  end)
  it('errors out when %b modifier is used for a float', function()
    eq('Vim(call):E805:Using a Float as a Number', exc_exec('call printf("%b", [])'))
  end)
  it('works with negative numbers when using %b', function()
    eq('…', funcs.printf('%b', -1))
  end)
end)
```

And note when writing tests: if your function produces exceptions or has any other corner cases (like negative numbers), writing tests cases with such functionality tends to have has more value then writing 100500 tests for various prefixes. Though in any case four test cases can hardly be considered enough extensive testing.

Test with all modifiers uses behaviour observed with %x, I have not actually checked what this PR does. Last test needs to contain actual value.

**Comment by uxcn** [comment]
Awesome, thank you again @ZyX-I.

> Regarding `long long`: if it was not `printf` I would suggest using `uint64_t`.

Normally I prefer exact width integers.

> By the way, I find this rather useless. Unlike hexadecimal there are no letter-digits.

I would agree, but symmetry is intuitive for people.  I'd rather add something harmless and not have anyone use it, than have someone expect it and not have it.

> And note when writing tests: if your function produces exceptions or has any other corner cases (like negative numbers), writing tests cases with such functionality tends to have has more value then writing 100500 tests for various prefixes.

Agreed.

Also FYI, I checked that Vim [patch](https://github.com/vim/vim/commit/887c1fea4a114e7170091942d0446c8882701b5b) for this, and the function signature for `vim_str2nr` was changed (bitmask).  I can update this patch to match, but I am thinking it might change again in the future (e.g. arbitrary precision).

I'll be pushing another update hopefully later tonight.

**Comment by uxcn** [comment]
I've updated this PR including the lint fixes and non-legacy test for `printf`.

The `vim_str2nr` changes made in the Vim version have been left out, at least for now.

**Comment by ZyX-I** [comment]
@uxcn Comment regarding ascii_isbdigit is still applicable.

File: src/nvim/charset.c [hunk 31]
```diff
@@ -1788,28 +1828,40 @@ void vim_str2nr(char_u *start, int *hexp, int *len, int dooct, int dohex,
...
-  } else if ((hex != 0) || (dohex > 1)) {
+  } else if (pre != 0 || dohex > 1) {
     // hex
+    if (pre != 0) {
+      n += 2;  // skip over "0x"
+    }
     while (ascii_isxdigit(*ptr)) {
       un = 16 * un + (unsigned long)hex2nr(*ptr);
-      ptr++;
+      ++ptr;
```
Comment (ZyX-I) [inline src/nvim/charset.c]: Revert.

File: src/nvim/ops.c [hunk 32]
```diff
@@ -4186,75 +4186,102 @@ static void reverse_line(char_u *s)
...
+///
+/// @return FAIL for failure, OK otherwise
 int do_addsub(int command, linenr_T Prenum1)
 {
   int col;
   char_u      *buf1;
   char_u buf2[NUMBUFLEN];
-  int hex;                      /* 'X' or 'x': hex; '0': octal */
-  static int hexupper = FALSE;          /* 0xABC */
+  int pre;                      // 'X' or 'x': hex; '0': octal; 'B' or 'b': bin
```
Comment (ZyX-I) [inline src/nvim/ops.c]: This comment is too shifted, but it is not aligned with anything.

File: src/nvim/ops.c [hunk 33]
```diff
@@ -4186,75 +4186,102 @@ static void reverse_line(char_u *s)
...
+
+  if (dobin) {
+    while (col > 0 && ascii_isbdigit(ptr[col])) {
+      --col;
+    }
+  }
+
+  if (dohex) {
+    while (col > 0 && ascii_isxdigit(ptr[col])) {
+      --col;
```
Comment (ZyX-I) [inline src/nvim/ops.c]: Here and above: `col--`, not `--col`.

File: src/nvim/ops.c [hunk 34]
```diff
@@ -4186,75 +4186,102 @@ static void reverse_line(char_u *s)
...
+       && ptr[col - 1] == '0'
+       && ascii_isxdigit(ptr[col + 1])) ||
+      (dobin
+       && col > 0
+       && (ptr[col] == 'B'
+           || ptr[col] == 'b')
+       && ptr[col - 1] == '0'
+       && ascii_isbdigit(ptr[col + 1]))) {
+       // Found hexadecimal or binary number, move to its start.
       --col;
```
Comment (ZyX-I) [inline src/nvim/ops.c]: As well as here.

File: src/nvim/options.lua [hunk 35]
```diff
@@ -1599,7 +1599,7 @@ return {
       deny_duplicates=true,
       alloced=true,
       varname='p_nf',
-      defaults={if_true={vi="octal,hex", vim="hex"}}
+      defaults={if_true={vi="bin,octal,hex", vim="bin,hex"}}
```
Comment (ZyX-I) [inline src/nvim/options.lua]: Bram did accept this vi default? Actually it is incorrect: vi default meant to be compatible with old vi and this is new functionality, so it by no means can appear there.

Comment (ZyX-I) [inline src/nvim/options.lua]: Though `'nrformats'` is marked as “not in vi”, as well as `CTRL-A` and `CTRL-X`, so this does not matter.

Comment (ZyX-I) [inline src/nvim/options.lua]: Guess then `vi` default means `vim` default and `vim` default is `neovim` default.

**Comment by ZyX-I** [comment]
Except for a few issues with pre-/postincrement, documentation and ascii_isbdigit LGTM. All of them should still not be hidden by “outdated diff”.

File: src/nvim/options.lua (hunk 35 above)
Comment (uxcn) [inline src/nvim/options.lua]: > Guess then vi default means vim default and vim default is neovim default, in which case this is completely correct.

This was what I guessed too.  The Vim code only has one set of defaults which are now `bin,octal,hex`.

**Comment by uxcn** [comment]
@ZyX-I I unfortunately rebased and force pushed these two commits since I added `printf` and the branch fell out of sync.  I ended up dropping the original lint fixes and re-coding instead of merging.  I went through and re-checked against the hidden diffs in the latest push though.

I try to keep the commits to as few and narrowly scoped as possible in general (e.g. one for feature and one for lint).  The idea is that it keeps the history clean and easier to work with.  My hope is that breaking it into two separate commits makes it easier to review the code changes distinct from the lint fixes.  I'm okay with whatever is easiest for everyone else though.

**Comment by watiko** [comment]
@uxcn Nice work! BTW Bram introduced a `what` flag and is there reason to use `dobin, dooct, dohec`?

**Comment by uxcn** [comment]
@watiko thanks :).  Yep, Bram changed the three psuedo-bools into a bitmask for `vim_str2nr` (`call_vim_function`, `ex_sort`, etc...), which is similar to what I was thinking about doing but I didn't want to commit to right away.  I was kind of hoping to get feedback from others first.

@ZyX-I also mentioned the issue with `<c-a>` and `<c-x>` not working with large values (overflow), which I've been keeping in the back of my mind.  I'd personally like to see support for arbitrary precision in general since there are a number of places where native types overflow (`<c-a>`, `<c-x>`, `:sort`, `<c-r>=2147483648`, etc...), but I think that could be a *much* bigger change.

**Comment by ZyX-I** [comment]
@uxcn Removing overflow in `<C-r>={expr}` is much bigger change because you need to change numbers to some kind of big integers and this will touch every single place where typval_T or varnumber_T numbers are used. Making `<C-a>` and `<C-x>` not overflow does not require this, you need to change just one function and a few places: basically vim_str2nr call and where value is converted back.

**Comment by uxcn** [comment]
@ZyX-I That's kind of what I was thinking.  Arbitrary precision in general would mean changing *everything* in VimL that understands numeric values.  Aside from the bugfix, I can think of a *lot* of other use cases though (IPv6 addresses, SIMD values, math in the expression register, non-trivial numbers in general, etc...).

There are definitely some easy wins before hopefully getting to that, but I think the plus could be that it simplifies/fixes a lot of other stuff.

**Comment by justinmk** [comment]
@uxcn Thanks for this! I fixed a small issue with the printf test: https://github.com/neovim/neovim/commit/542c79f20d7835951e56a3c9f86c9c7ee9b0a531

**Comment by uxcn** [comment]
@justinmk No problem, I appreciate helping me merge it.

Do you mind if I give a patch for the `printf` code to Bram?

**Comment by justinmk** [comment]
@uxcn please do. You own the copyright, of course :)
